File: CS/E1520/WinSolution.Module.Win/VGridControlDetailViewItem.cs
using System;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp;
using DevExpress.XtraVerticalGrid;
using DevExpress.Data.Filtering;
using DevExpress.Xpo;
using DevExpress.ExpressApp.NodeWrappers;
using System.Collections;
using DevExpress.XtraEditors.Repository;
using DevExpress.ExpressApp.Win.Editors;
using DevExpress.XtraVerticalGrid.Rows;
using DevExpress.ExpressApp.DC;
using DevExpress.ExpressApp.Model;

namespace WinSolution.Module.Win {
	public interface IModelVGridControlDetailViewItem : IModelDetailViewItem {
	}
	[DetailViewItemAttribute(typeof(IModelVGridControlDetailViewItem))]
    public class VGridControlDetailViewItem : ViewItem, IComplexPropertyEditor {
		public VGridControlDetailViewItem(IModelVGridControlDetailViewItem model, Type classType) : base(classType, model.Id) { }
        private VGridControl vGridControlCore = null;
        public VGridControl VGridControl {
            get {
                return vGridControlCore;
            }
        }
        private XafApplication applicationCore = null;
        public XafApplication Application {
            get {
                return applicationCore;
            }
        }
        protected override object CreateControlCore() {
            vGridControlCore = new VGridControl();
            vGridControlCore.DataSource = GetDataSource();
            RepositoryEditorsFactory factory = new RepositoryEditorsFactory(Application, View.ObjectSpace);
			IModelDetailView dw = (IModelDetailView)View.Model;
            foreach (IModelDetailViewItem item in dw.Items) {
				IModelPropertyEditor modelPropertyEditor = item as IModelPropertyEditor;
				if(modelPropertyEditor != null) {
					bool isGranted = DataManipulationRight.CanRead(ObjectType, modelPropertyEditor.PropertyName, View.CurrentObject, null);
					RepositoryItem repositoryItem = factory.CreateRepositoryItem(!isGranted, modelPropertyEditor, ObjectType);
					if(repositoryItem != null) {
						IMemberInfo mi = ObjectTypeInfo.FindMember(modelPropertyEditor.PropertyName);
						if(mi != null) {
							vGridControlCore.RepositoryItems.Add(repositoryItem);
							EditorRow row = new EditorRow(mi.BindingName);
							row.Properties.Caption = item.Caption;
							row.Properties.RowEdit = repositoryItem;
							vGridControlCore.Rows.Add(row);
						}
					}
				}
            }
            vGridControlCore.CellValueChanged += OnVGridControlCellValueChanged;
            View.CurrentObjectChanged += OnViewCurrentObjectChanged;
            View.ObjectSpace.Reloaded += OnObjectSpaceReloaded;
            return vGridControlCore;
        }
        protected override void Dispose(bool disposing) {
            if (disposing && (vGridControlCore != null) && (View != null)) {
                View.CurrentObjectChanged -= OnViewCurrentObjectChanged;
                vGridControlCore.CellValueChanged -= OnVGridControlCellValueChanged;
                View.ObjectSpace.Reloaded -= OnObjectSpaceReloaded;
            }
            base.Dispose(disposing);
        }
        private CriteriaOperator GetCriteria() {
            if (View.CurrentObject != null) {
                return CriteriaOperator.Parse(string.Format("{0}=?", View.ObjectSpace.GetKeyPropertyName(ObjectType)), View.ObjectSpace.GetKeyValue(View.CurrentObject));
            } else {
                return null;
            }
        }
        private IList GetDataSource() {
            XPCollection ds = new XPCollection(View.ObjectSpace.Session, ObjectType, false);
            ds.Add(View.CurrentObject);
            return ds;
        }
        private void OnViewCurrentObjectChanged(object sender, EventArgs e) {
            VGridControl.DataSource = GetDataSource();
        }
        private void OnObjectSpaceReloaded(object sender, EventArgs e) {
            VGridControl.DataSource = GetDataSource();
        }
        private void OnVGridControlCellValueChanged(object sender, DevExpress.XtraVerticalGrid.Events.CellValueChangedEventArgs e) {
            VGridControl.BindingContext[View.CurrentObject].EndCurrentEdit();
        }
        #region IComplexPropertyEditor Members
        public void Setup(ObjectSpace objectSpace, XafApplication application) {
            applicationCore = application;
        }
        #endregion
    }
}

File: VB/WinSolution.Module.Win/VGridControlDetailViewItem.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp
Imports DevExpress.XtraVerticalGrid
Imports DevExpress.Data.Filtering
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp.NodeWrappers
Imports System.Collections
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.ExpressApp.Win.Editors
Imports DevExpress.XtraVerticalGrid.Rows
Imports DevExpress.ExpressApp.DC

Namespace WinSolution.Module.Win
	<DetailViewItemName("VGridControlDetailViewItem")> _
	Public Class VGridControlDetailViewItem
		Inherits DetailViewItem
		Implements IComplexPropertyEditor
		Public Sub New(ByVal classType As Type, ByVal info As DictionaryNode)
			MyBase.New(info, classType)
		End Sub
		Private vGridControlCore As VGridControl = Nothing
		Public ReadOnly Property VGridControl() As VGridControl
			Get
				Return vGridControlCore
			End Get
		End Property
		Private applicationCore As XafApplication = Nothing
		Public ReadOnly Property Application() As XafApplication
			Get
				Return applicationCore
			End Get
		End Property
		Protected Overrides Function CreateControlCore() As Object
			vGridControlCore = New VGridControl()
			vGridControlCore.DataSource = GetDataSource()
			Dim factory As New RepositoryEditorsFactory(Application, View.ObjectSpace)
			Dim dw As New DetailViewInfoNodeWrapper(View.Info)
			For Each item As DetailViewItemInfoNodeWrapper In dw.Editors.Items
				Dim isGranted As Boolean = DataManipulationRight.CanRead(ObjectType, item.PropertyName, View.CurrentObject, Nothing)
				Dim repositoryItem As RepositoryItem = factory.CreateRepositoryItem((Not isGranted), item, ObjectType)
				If repositoryItem IsNot Nothing Then
					Dim mi As IMemberInfo = ObjectTypeInfo.FindMember(item.PropertyName)
					If mi IsNot Nothing Then
						vGridControlCore.RepositoryItems.Add(repositoryItem)
						Dim row As New EditorRow(mi.BindingName)
						row.Properties.Caption = item.Caption
						row.Properties.RowEdit = repositoryItem
						vGridControlCore.Rows.Add(row)
					End If
				End If
			Next item
			AddHandler vGridControlCore.CellValueChanged, AddressOf OnVGridControlCellValueChanged
			AddHandler View.CurrentObjectChanged, AddressOf OnViewCurrentObjectChanged
			AddHandler View.ObjectSpace.Reloaded, AddressOf OnObjectSpaceReloaded
			Return vGridControlCore
		End Function
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing AndAlso (vGridControlCore IsNot Nothing) AndAlso (View IsNot Nothing) Then
				RemoveHandler View.CurrentObjectChanged, AddressOf OnViewCurrentObjectChanged
				RemoveHandler vGridControlCore.CellValueChanged, AddressOf OnVGridControlCellValueChanged
				RemoveHandler View.ObjectSpace.Reloaded, AddressOf OnObjectSpaceReloaded
			End If
			MyBase.Dispose(disposing)
		End Sub
		Private Function GetCriteria() As CriteriaOperator
			If View.CurrentObject IsNot Nothing Then
				Return CriteriaOperator.Parse(String.Format("{0}=?", View.ObjectSpace.GetKeyPropertyName(ObjectType)), View.ObjectSpace.GetKeyValue(View.CurrentObject))
			Else
				Return Nothing
			End If
		End Function
		Private Function GetDataSource() As IList
			Dim ds As New XPCollection(View.ObjectSpace.Session, ObjectType, False)
			ds.Add(View.CurrentObject)
			Return ds
		End Function
		Private Sub OnViewCurrentObjectChanged(ByVal sender As Object, ByVal e As EventArgs)
			VGridControl.DataSource = GetDataSource()
		End Sub
		Private Sub OnObjectSpaceReloaded(ByVal sender As Object, ByVal e As EventArgs)
			VGridControl.DataSource = GetDataSource()
		End Sub
		Private Sub OnVGridControlCellValueChanged(ByVal sender As Object, ByVal e As DevExpress.XtraVerticalGrid.Events.CellValueChangedEventArgs)
			VGridControl.BindingContext(View.CurrentObject).EndCurrentEdit()
		End Sub
		#Region "IComplexPropertyEditor Members"
		Public Sub Setup(ByVal objectSpace As ObjectSpace, ByVal application As XafApplication) Implements IComplexPropertyEditor.Setup
			applicationCore = application
		End Sub
		#End Region
	End Class
End Namespace